src/Workspaces/Core/Portable/Utilities/AsyncLazy`1.cs

```diff
@@ -294,6 +294,13 @@ private Request CreateNewRequest_NoLock()
 
         public override Task<T> GetValueAsync(CancellationToken cancellationToken)
```

Comment:
There is an argument we should cancel. Imagine you have some pattern like:
```
var tree = await document.GetTreeAsync(cancellationToken);
var semanticModel = await document.GetSemanticModelAsync(cancellationToken);

foreach (var symbol in GetSomeSymbolsFromSemanticModel(semanticModel))
{
    cancellationToken.ThrowIfCancellationRequested();
    // do stuff with symbol
}
```

if cancellation happened before the beginning of this snippet, we're not going to observe the cancellation until much later after we've potentially wasted more time. If we observe it sooner, then we're not churning extra CPU doing work that is just going to give up later.

The question is "what is the overhead of CPU doing the check vs. the wasted CPU that is only going to observe the cancellation later?" My gut is saying do the check, the assumption being that checking the cancellation token is a handful of x86 instructions, and resuming a task on the thread pool only to immediately observe the cancellation token is thousands.